### PR TITLE
Add wave 2dvar capability

### LINKED ISSUE #610: Add wave 2dvar capability
## Description
To include wave component (significant wave height in this case) in the SOCA 3DVAR ctest. Here we start simple by only assimilating sea surface significant wave height using the Identity operator. 

## Solution
- Add a sample swh buoy obs file
- Add a sample wave bkgerr file
- Add swh variable in fields_metadata.yml
- Modify soca_fields_mod.F90 to add capability to read wave background    
- Modify  static_socaerror_init.yml to add wave-related b matrix
- Modify 3dvar_soca.yml to add wave component
- Update 3dvar_soca testref file

### PR DESCRIPTION
**Pull request by clouden90**
## Description

To include wave component in the SOCA. Here we start simple by only assimilating swh (sea surface significant wave height ) using the Identity operator.

## What have been done in this PR
- Add a sample swh obs file
- Add a sample wave bkgerr file
- Add swh variable in fields_metadata.yml
- Modify soca_fields_mod.F90 to add capability to read wave background
- Modify soca_bkgerrutil_mod.F90, soca_bkgerrgodas_mod.F90, and soca_bkgerr_mod.F90 for specification of wave background error.
- Modify static_socaerror_init.yml to add wave bump parametric b
- Modify 3dvar_soca.yml to add wave component
- Update 3dvar_soca testref file

### Issue(s) addressed

- fixes #610

## Testing

Tested on both Hear and Orion with intel compiler. Additionally, a near real-time test case based on this PR was conducted, and detailed results  can be found [here](https://drive.google.com/file/d/1AoMRzLH5VRZTNHZzhrFFgjBN2-i1qoqv/view?usp=sharing) if you are interested.

### CODE REVIEW

File: src/soca/Transforms/BkgErrGodas/soca_bkgerrgodas_mod.F90 [hunk 1]
```diff
@@ -78,12 +78,12 @@ subroutine soca_bkgerrgodas_setup(f_conf, self, bkg, geom)
...
   call soca_bkgerrgodas_ssh(self)
 
-  ! Invent background error for ocnsfc and ocn_bgc fields: set 
+  ! Invent background error for ocnsfc, wav and ocn_bgc fields: set 
   ! it to 10% or 20% of the background for now ...
   do i=1,size(self%std_bkgerr%fields)
     field => self%std_bkgerr%fields(i)
     select case(field%name)
-    case ('sw','lw','lhf','shf','us')
+    case ('sw','lw','lhf','shf','us','wsh')
```
Comment (guillaumevernieres) [inline src/soca/Transforms/BkgErrGodas/soca_bkgerrgodas_mod.F90]: Typo swh, not wsh ... A flaw in our ci for not picking that up!

File: src/soca/Transforms/BkgErr/soca_bkgerr_mod.F90 [hunk 2]
```diff
@@ -79,14 +79,14 @@ subroutine soca_bkgerr_setup(f_conf, self, bkg, geom)
...
-  ! Invent background error for ocnsfc and ocn_bgc fields: 
+  ! Invent background error for ocnsfc, wav and ocn_bgc fields: 
   ! set it to 10% or 20% of the background for now ...
   ! TODO: Read background error for ocnsfc and ocn_bgc from 
   ! files
   do i=1,size(self%std_bkgerr%fields)
     field => self%std_bkgerr%fields(i)
     select case(field%name)
-    case ('sw','lw','lhf','shf','us')
+    case ('sw','lw','lhf','shf','us','swh')
```
Comment (guillaumevernieres) [inline src/soca/Transforms/BkgErr/soca_bkgerr_mod.F90]: ~~Will you remember to remove `swh`from the case when you have a more realistic background error?~~

Edit: I see that you added a `wav_bkgerror.nc` which presumably contains the variance of the background error for swh, the above is overwriting it. You should remove swh from the case statement.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -150,6 +152,7 @@ set( soca_obs
   Data/Obs/prof.nc
   Data/Obs/sss.nc
   Data/Obs/sst.nc
+  Data/Obs/hs.nc
```
Comment (guillaumevernieres) [inline test/CMakeLists.txt]: pardon my ignorance, but what does `hs` stand for?

**Review by guillaumevernieres — CHANGES_REQUESTED** [review]
Thanks for getting that project off the ground @clouden90.

I know this wasn't done for the bgc variables, but you should add `swh` to the `getvalue` test. I have a few minor comments but otherwise it looks fine to me.

File: src/soca/Transforms/BkgErrGodas/soca_bkgerrgodas_mod.F90 (hunk 1 above)
Comment (clouden90) [reply src/soca/Transforms/BkgErrGodas/soca_bkgerrgodas_mod.F90]: Good catch! Always good to have another eyes. Fixed

File: src/soca/Transforms/BkgErr/soca_bkgerr_mod.F90 (hunk 2 above)
Comment (clouden90) [reply src/soca/Transforms/BkgErr/soca_bkgerr_mod.F90]: Removed

File: test/Data/fields_metadata.yml [hunk 4]
```diff
@@ -80,6 +80,15 @@
   io name: hsnon
   property: positive_definite
 
+# --------------------------------------------------------------------------------------------------
+# wave state variables
+# --------------------------------------------------------------------------------------------------
+- name: swh
+  getval name: sea_surface_significant_wave_height
```
Comment (travissluka) [inline test/Data/fields_metadata.yml]: We are trying to use CF standard names wherever possible
(https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html)
can you switch this to `sea_surface_wave_significant_height` ?

**Review by travissluka — CHANGES_REQUESTED** [review]
Thanks @clouden90 !
One small comment on the variable name.

File: test/CMakeLists.txt (hunk 3 above)
Comment (clouden90) [reply test/CMakeLists.txt]: hs = significant height of combined wind waves and swell (m). It is quite a standard variable name used in ww3 as well as field of wind wave modeling/observation. I can rename it to swh if you think it is more appropriate in our case.

File: test/Data/fields_metadata.yml (hunk 4 above)
Comment (clouden90) [reply test/Data/fields_metadata.yml]: No problem, will change it.

File: test/CMakeLists.txt (hunk 3 above)
Comment (guillaumevernieres) [reply test/CMakeLists.txt]: shouldn't it be `sh` then? I'm not going to block that pr for the naming of a file, just curious of the logic of the naming and if that's really what you wanted.

**Comment by clouden90** [comment]
@guillaumevernieres @travissluka , Thx for all the comments! I have done the changes based on most of your comments. Additionally, `swh`  is added to the `getvalue` test. This PR is ready, just wait for the Travis CI check, but I think it should be ok.

**Review by travissluka — APPROVED** [review]
:ocean: :+1:

**Review by guillaumevernieres — APPROVED** [review]
:+1:  :tada: :tada: :tada:

**Pull request merged**